CIP-????/README.md

```diff
@@ -0,0 +1,168 @@
+
+``` Haskell
+Nothing = Constr 0 []
+Just datum = Constr 1 [datumAsData]
+```
+
+The datum will only be converted from `BuiltinData` if the `Just` is going to be used. No parsing should be attempted if the `Nothing` is going to be used.
+
+As an example, to pass the unit datum into a plutus spending script, the following `Data` would be passed:
+``` Haskell
+Just () = Constr 1 [Constr 0 []]
+```
+
+### Creating Universal Plutus Scripts
+All plutus scripts will now be written as:
+```
+mkUniversalScript :: Maybe Datum -> Redeemer -> ScriptContext -> Bool
+mkUniversalScript (Just d) r ctx = ...
+mkUniversalScript Nothing r ctx = ...
+```
+
+### Using Universal Plutus Scripts
+#### Minting and Staking
+When a universal script is to be used for either minting or staking, `Nothing` will automatically be passed into the script. A `plutus-script-v3` flag will be needed to differentiate whether a universal script is being used or a v1/v2 script is being used.
+
+#### Spending
+Using the universal scripts, there will now be two ways to use a spending script.
+
+1. **UTxO level spending scripts**: These scripts will have the datum attached to the UTxO being spent parsed and passed to the universal script as `Just datum`. These scripts are meant to be executed once per UTxO consumed from the script address.
+2. **Transaction level spending scripts**: These scripts will have `Nothing` passed to the universal script without attempting to parse the datums attached to the UTxOs being spent. These scripts are meant to be executed once per transaction. The datums will still be present in the `ScriptContext`.
+
+#### `cardano-cli`
+`cardano-cli` will need to be expanded to add `--utxo-level` and `--tx-level` flags so that users can explicitly say which level of the spending script to use.
+
+Example UTxO level spending script flag usage:
+``` Bash
+  --tx-in <script_utxo_to_spend> \
+  --utxo-level-spending-tx-in-reference <ref_script_utxo_id> \
+  --plutus-script-v3 \
+  --utxo-level-spending-reference-tx-in-inline-datum-present \
+  --utxo-level-spending-reference-tx-in-redeemer-file <script_redeemer> \
+```
+
+Example transaction level spending script flag usage:
+``` Bash
+  --tx-in <script_utxo1> \
+  --tx-in <script_utxo2> \
+  --tx-in <script_utxo3> \
+  --tx-level-spending-tx-in-reference <ref_script_utxo_id> \
+  --plutus-script-v3 \
+  --tx-level-spending-reference-tx-in-redeemer-file <script_redeemer> \
+```
+
+For transaction level scripts, since there is no need to parse the datums attached to each UTxO, the datum flags are not needed. Any datums present will still appear in the `ScriptContext` that gets passed to the plutus script. All UTxOs being validated by the transaction level script are assumed to be using the same redeemer which is why it is okay to place the redeemer flag at the end.
```

Comment:
What if we're building a transaction involving multiple tx-level scripts each expecting their own redeemer data (not compatible with others)?

Reply:
Each tx-level script will get its own redeemer. So if Script A and Script B were used in the same tx, Script A will use redeemer A and Script B will use redeemer B. The `--tx-level` redeemer flag is meant to be paired up with a specific `--tx-level` script for this reason. This proposal does not make all scripts in the tx use the same redeemer.